src/line_launcher/mod.rs

```diff
@@ -62,6 +68,70 @@ impl<'progression> ProgressionState<'progression> {
 const CHANNEL: Channel = Channel::Ch1;
 const VELOCITY: u8 = 100;
 
+struct NoteOffInstruction {
+    note: Note,
+    time: SystemTime,
+}
+
+struct NoteOffTriggerer {
+    receiver: Receiver<NoteOffInstruction>,
+    midi_message_sender: Arc<Mutex<MidiMessageSender>>,
+    playing_state: Arc<Mutex<PlayingState>>,
+}
+
+impl NoteOffTriggerer {
+    pub fn new(
+        midi_message_sender: Arc<Mutex<MidiMessageSender>>,
+        playing_state: Arc<Mutex<PlayingState>>,
+    ) -> (Self, Sender<NoteOffInstruction>) {
+        let (sender, receiver) = mpsc::channel();
+
+        let note_off_triggerer = Self {
+            receiver,
+            midi_message_sender,
+            playing_state,
+        };
+
+        (note_off_triggerer, sender)
+    }
+
+    pub fn listen(self) -> JoinHandle<()> {
+        thread::spawn(move || loop {
+            let note_off_instruction = self.receiver.recv().unwrap();
+
+            let now = SystemTime::now();
+            let from_now = note_off_instruction
+                .time
+                .duration_since(now)
+                .unwrap_or_default();
+
+            spin_sleep::sleep(from_now);
+
+            let mut playing_state = self.playing_state.lock().unwrap();
+
+            if let PlayingState::Playing {
+                has_fired_previous_note_off: false,
```

Comment:
In theory this is coordinating where the existing `possibly_trigger_notes()` method will trigger the existing behavior (where it'll turn the last note off when turning the next note on) if for some reason it hasn't been turned off yet. But then for this to really work, I think this `NoteOffTriggerer` would need to be able to verify that the current playing state is the same note it was intended to turn off (probably it needs to be sent `note_index` inside `NoteOffInstruction` or something like that) so that it wouldn't turn off _following_ notes if it's running after the next note on has happened